oidc.go

```diff
@@ -0,0 +1,196 @@
+package headscale
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/gin-gonic/gin"
+	"github.com/patrickmn/go-cache"
+	"github.com/rs/zerolog/log"
+	"golang.org/x/oauth2"
+	"gorm.io/gorm"
+	"net/http"
+	"time"
+)
+
+type IDTokenClaims struct {
+	Name     string   `json:"name,omitempty"`
+	Groups   []string `json:"groups,omitempty"`
+	Email    string   `json:"email"`
+	Username string   `json:"preferred_username,omitempty"`
+}
+
+var oidcProvider *oidc.Provider
+var oauth2Config *oauth2.Config
+var stateCache *cache.Cache
+
+// RegisterOIDC redirects to the OIDC provider for authentication
+// Puts machine key in cache so the callback can retrieve it using the oidc state param
+// Listens in /oidc/register/:mKey
+func (h *Headscale) RegisterOIDC(c *gin.Context) {
+	mKeyStr := c.Param("mkey")
+	if mKeyStr == "" {
+		c.String(http.StatusBadRequest, "Wrong params")
+		return
+	}
+
+	var err error
+
+	// grab oidc config if it hasn't been already
+	if oauth2Config == nil {
+		oidcProvider, err = oidc.NewProvider(context.Background(), h.cfg.OIDCIssuer)
+
+		if err != nil {
+			log.Error().Msgf("Could not retrieve OIDC Config: %s", err.Error())
+			c.String(http.StatusInternalServerError, "Could not retrieve OIDC Config")
+			return
+		}
+
+		oauth2Config = &oauth2.Config{
+			ClientID:     h.cfg.OIDCClientID,
+			ClientSecret: h.cfg.OIDCClientSecret,
+			Endpoint:     oidcProvider.Endpoint(),
+			RedirectURL:  fmt.Sprintf("%s/oidc/callback", h.cfg.ServerURL),
+			Scopes:       []string{oidc.ScopeOpenID, "profile", "email"},
+		}
+
+	}
```

Comment:
This section looks like it can have the potential for a race condition, as per comment about moving it from global to the `Headscale` struct. Can this be instantiated once at startup so we are sure we dont end up there?